common/uid_store.c

```diff
@@ -17,22 +17,72 @@
  * Copyright (C) 2015 Simon Newton
  */
 
-#include "uid_store.h"
 #include "common_settings.h"
+#include "config_options.h"
+#include "uid_store.h"
 
 #include <stdint.h>
 
-#ifdef CFG_USE_DEVELOPMENT_UID
-// Fallback to a static development UID
-static const uint8_t test_uid[UID_LENGTH] = {
-  0x7a, 0x70, 0xff, 0xff, 0xff, 0x00
+#ifndef CFG_UID_SOURCE
+#error "CFG_UID_SOURCE is not set"
+#endif
+
+#if CFG_UID_SOURCE == UID_FROM_MAC
+#include "peripheral/eth/plib_eth.h"
+
+static uint8_t kUIDArray[UID_LENGTH];
+
+inline uint8_t ShiftRight(uint8_t b) {
+  return (b >> 4) & 0x0f;
+}
+
+inline uint8_t ShiftLeft(uint8_t b) {
+  return (b << 4) & 0xf0;
+}
+
+void UIDStore_Init() {
+  // The UID is derived from the RDM manufacturer ID & the MAC address.
+  // The first 3 bytes of the MAC address is the Microchip OID 00:1e:c0 which
+  // is constant. The bottom 3 bytes contain the unique serial number.
+  //
+  // To support more than one responder per device, we set the lower 4 bits of
+  // the UID to 0 so we have 16 responders per device. These means the complete
+  // UID takes the form:
+  //   MMMM:0AAAAAA0
+  // Where M is the PLASA manufacturer ID and A are the values from the MAC
+  // address.
+  kUIDArray[0] = CFG_MANUFACTURER_ID >> 8;
+  kUIDArray[1] = CFG_MANUFACTURER_ID & 0xff;
+  kUIDArray[2] = ShiftRight(PLIB_ETH_StationAddressGet(ETH_ID_0, 4));
+  kUIDArray[3] = ShiftLeft(PLIB_ETH_StationAddressGet(ETH_ID_0, 4)) +
+                 ShiftRight(PLIB_ETH_StationAddressGet(ETH_ID_0, 5));
+  kUIDArray[4] = ShiftLeft(PLIB_ETH_StationAddressGet(ETH_ID_0, 5)) +
+                 ShiftRight(PLIB_ETH_StationAddressGet(ETH_ID_0, 6));
+  kUIDArray[5] = ShiftLeft(PLIB_ETH_StationAddressGet(ETH_ID_0, 6));
+}
+
+const uint8_t* const kUID = &kUIDArray[0];
+
+#elif CFG_UID_SOURCE == UID_FROM_FLASH
+// Read from the magic linker symbol.
+extern uint8_t __attribute__((space(prog))) _uid;
+const uint8_t* const kUID = &_uid;
+
+void UIDStore_Init() {}
+
+#elif CFG_UID_SOURCE == UID_TEST_UID
+
+static const uint8_t kDevUID[UID_LENGTH] = {
+  CFG_MANUFACTURER_ID >> 8,
+  CFG_MANUFACTURER_ID & 0xff,
```

Comment:
This is currently the first OLA dummy device: https://wiki.openlighting.org/index.php/Open_Lighting_Allocations#UIDs

Why not bump it up to 7a70:fffffffe or into the developement range as 7a70:fffffe00?